### DISCUSSION TITLE
RFC: Support Markdown images relative path

### DISCUSSION THREAD

**Discussion by wassfila** [withastro/roadmap#376] [Proposal]
# Proposal
Add support for Astro markdown to reference images relative to the markdown file

usage examples :

```markdown
![Tree](./tree.svg)

![Github](github-dark.png)
```

## Motivation
This proposal is to make markdown portable for programmers and easy to use for non programmers.

I like Astro and believe this will increase its adoption by avoiding rework when using existing markdown, this will also improve DX by having markdown rendered in code editor same as by Astro.

- This helps keep images in the same directory as the markdown using them, and not requiring the markdown author to understand the build system
- This is a defacto standard because it already works out of the box in VS Code and Github

# Survey
- core image optimization rfc https://github.com/withastro/roadmap/discussions/468  : this rfc has partial but not total coverage due to missing SVG support, and in case optimization cannot be disabled in favor of usage of simple assets management alone. Also might prevent customizing images with a custom component.

- astro doc
https://docs.astro.build/en/guides/markdown-content/

- astro mdx integration
https://docs.astro.build/en/guides/integrations-guide/mdx/

- astro examples
https://github.com/withastro/astro/tree/main/examples

- active RFCs proposals :
https://github.com/withastro/rfcs/tree/main/proposals

- open RFC's PR's : 
https://github.com/withastro/rfcs/pulls

  - https://github.com/withastro/rfcs/pull/381

- Astro image components
https://docs.astro.build/en/guides/integrations-guide/image/

# Standard
Why I mention that this a defacto standard and not a standard because searching in common cheat sheets, I could not find a clear reference to relative images usage

https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#images

This link show a relative image on the summary table then when following the link only absolute paths are used
https://www.markdownguide.org/cheat-sheet/

# Defacto Standard
## usage in VSCode
in VSCode the markdown preview of relative file usage already renders the expected result
<img width="683" alt="image" src="https://user-images.githubusercontent.com/1510763/200125288-9aef0c6f-1c8c-42d3-b206-b8ee0ecc3627.png">

## usage in github
in github, such a committed markdown file with the asset in the same directory will also render, 
![image](https://user-images.githubusercontent.com/1510763/200125685-73e9cd16-677d-4483-a691-be3a85e3afe5.png)

a closer look by inspecting the html source of the github page shows management of the assets in a raw directory
```
<img src="/MicroWebStacks/astro-big-doc/raw/main/src/pages/github-dark.png" alt="github dark" style="max-width: 100%;">
```
# Solution with integration
https://github.com/userquin/astro-md-image-integration
by @userquin and @kasvith 

# Solution implemented with remark plugin
(note : this solution did not take existence of Astro image integration in consideration which is currently ongoing)
I implemented a solution that solves the problem by managing the asset replacing the relative path with the new one of the copied asset in the public folder, this way, it is relying on Astro to move the assets further from the public directory to the final distribution folder. Final solution should avoid that.

## how this works
This is similar to assets management in the src dir by Astro but extends it to the root directory of the project where markdown can also be placed in e.g. `data` directory. All managed assets will get a unique path und the `public/raw` directory. The path relative to the root will be reflected from the `raw` which is like the new asset root.

So all the plugin does is copy the asset to the new `public/raw` path and update the node.url so that the html sees the new asset path. given that all paths under the root are unique, no need to rename assets file names with generated uids.

## code

Link : https://github.com/MicroWebStacks/astro-big-doc/blob/main/src/libs/remark-rel-asset.js

This is the full remark-rel-asset.js plugin
```javascript
import {visit} from "unist-util-visit";
import {dirname} from 'path'
import {relAssetToUrl} from './utils'
import config from '../../astro.config.mjs'

function remarkRelAsset() {
  return function transformer(syntaxTree,file) {
    console.log(` 'remarkRelAsset' * in file '${file.history}'`)
    visit(syntaxTree,  node => {
      if(!['image'].includes(node.type)){
        return
      }
      node.url = relAssetToUrl(node.url,dirname(file.history[0]),"/"+config.base+"/")
    });
    return syntaxTree;
  };
}

export{
  remarkRelAsset
}
```

And this is where the magic happens 

Link : https://github.com/MicroWebStacks/astro-big-doc/blob/main/src/libs/utils.js

```javascript
import {existsSync,copyFileSync,mkdirSync} from 'fs'
import {resolve,normalize,dirname,join,relative} from 'path'
//import config from '../../astro.config.mjs'

function relAssetToUrl(relativepath,refdir,baseUrl){
    let newurl = relativepath
    const filepath = join(refdir,relativepath)
    if(existsSync(filepath)){
      const rootdir = join(dirname(normalize(import.meta.url)),"../..").replace("file:\\","")
      const targetroot = join(rootdir,"public/raw")
      const filerootrel = relative(rootdir,refdir)
      const targetpath = resolve(targetroot,filerootrel)
      const targetfile = join(targetpath,relativepath)
      const targetdir = dirname(targetfile)
      //console.log(`copy from '${filepath}' to '${targetfile}'`)
      const newpath = join("raw/",filerootrel,relativepath)
      newurl = baseUrl+ newpath.replaceAll('\\','/')
      //console.log(`  * new asset url = '${newurl}'`)
      if(!existsSync(targetdir)){
        mkdirSync(targetdir,{ recursive: true })
      }
      copyFileSync(filepath,targetfile)
    }

    return newurl
}

export{
    relAssetToUrl
}
```

## Caveats
* This solution is a quick proof of concept and has not been tested thoroughly
* referencing of files outside the root project folder might end up in unexpected behaviour
* This does not intend to generalize usage of imported markdown which might require additional security checks

## usage in Astro component
Although this was first intended to solve a standardized markdown relative paths notation, this can also be used in an Astro component as follows
```javascript
const newurl = relAssetToUrl(relFile,filedir,import.meta.env.BASE_URL)
```
here a link to a full real case usage example : https://github.com/MicroWebStacks/astro-big-doc/blob/c51ec72e9ee8e5e9683fbdecc19b16a4e27a67e4/src/components/gallery.astro#L19

but that is more subject to discussion, because first it relies on another remark plugin to provide the filedir, and there are plenty of alternatives like imports or Astro image integration. It still feels to me more flexible given the usage of build time dynamic variables and not having to hardcode all used files to an import.

# Expected Final solution
expected is out of the box Astro Markdown support for relative paths and relative assets management without intermediate public folder usage. This can relate to Astro image components integration, but goal is to have it applied directly to Markdown.

see also integation potenial ready to take in Astro https://github.com/userquin/astro-md-image-integration

**Comment by kasvith** [comment]
We also made an integration https://github.com/userquin/astro-md-image-integration

**Reply by wassfila** [reply]
That's really cool, that's the idea. Though in your examples, imgs paths are relative to the blog, not to the markdown file, so for example, they don't render when opened in github, that'd be for me an important cross compatibility criterium.

**Reply by kasvith** [reply]
They are working in github. Check https://github.com/userquin/astro-md-image-integration/blob/master/content/blog/first-post/post.md

It support

- Relative path to the post
- Images from the public directory(well if you are using)
- Can resolve any relative path and will included in final build

@userquin and me working on this still, will stabilize soon

**Reply by userquin** [reply]
> They are working in github. 

the absolute image (first one) not working !, the relative ones working... ;)

**Reply by kasvith** [reply]
First image refers to a image in Astro public directory, it works on dev and prod...not in GH

For your case, relative images always work

**Reply by wassfila** [reply]
I confirm, it is working.
The absolute path is of course not in scope as it is to show specific cases for Astro, but note that, the relative ones did not work on first attempt, which confused me, either slow loading due to server connection, but they only worked once I curiously reloaded the page again, and after that, they stay in the cache so work from first following load.

So indeed, I'm also looking closer to the getImage function from the astro image integration. I wish such things were part of Astro, but as an integration I think is acceptable as fallback.

**Reply by kasvith** [reply]
IMHO this integration should be defacto in Astro. I also faced the same issue as yours, even opened a thread like this in this RFCs. But this was the only solution we had at the end

**Reply by wassfila** [reply]
> IMHO this integration should be defacto in Astro. I also faced the same issue as yours, even opened a thread like this in this RFCs. But this was the only solution we had at the end

cool, I updated the description to mention your solution, thanks.

**Reply by kasvith** [reply]
@wassfila main author is @userquin :)

**Comment by kasvith** [comment]
We are on Astro2.0 and still this is not supported :|

**Reply by wassfila** [reply]
hmm, I went through the RFC Discussions around 250 Open discussions with my eye counting 29 with 6 votes and above.
I think one way to push this forward is to consider a PR Strategy without relying on an integration or plugin and consider impact on the ASTRO API. For that we need a deeper understanding of the Astro architecture, I'm not there yet myself. The good thing is that Content Collections was one of the major additions to Astro, so I think that this is high on the prio list, but we need closely watch other related roadmap work and try to link to this discussion.

**Reply by kasvith** [reply]
Had a discussion in discord, it seems they will work on this soon()

**Comment by andrezero** [comment]
Hey folks! 

I didn't find this thread when migrating my website 😢 (which was weird because I googled for 2 weeks from every possible angle how to get relative paths working for me)

## Why relative media paths are desirable (for me)

### 1) Portable content (framework agnostic)

As a content owner, I want to keep my MD(X) agnostic to framework opinions/constraints, so that I can port it easily

... I want to be able to preview content + media with most basic of MD readers
... I want to be able to migrate all my content to another website without having to do (too many) mods 

_(I work with design system documentation often, and this was a deal breaker to adopt Astro when we evaluated)_

### 2) low complexity (variance generates waste)

As a developer, I want to treat all images in the same way no matter where they're expressed in content or (meta)data, so that I can keep complexity level to a minimum

Treating all of these the same

- markdown `![](./media/image.jpg)` 
- component invocation `<Figure src="./media/image.jpg" ../>` 
- frontmatter field `thumbnail: ./media/image.jpg` 

... and even ...
- external data `[{"thing": "foo", "image": "./media/image.jpg}]`

... enables using the same approach (and basically the same code), whether you are rendering the image in the page, on a list, on metadata or even back to JSON.

_Note: images from external data need a base path context in order to be resolved, all other cases are simply resolved from the content (MD file) path_

## What I rolled out on my website

I pieced together an integration that does 
- relative images
- dev server
- image profiles
- local / remote images
- retrieving aspect ratio to eliminate CLS
- retrieving "dominant colour" (just a POC)

### How it works
- in dev / server - all images are rendered via endpoint [getImageEndpoint()](https://github.com/andrezero/andretorgal-site/blob/main/src/integration/images/utils/getImageEndpoint.ts)
- and in static build - all images via [getPublicPath](https://github.com/andrezero/andretorgal-site/blob/main/src/integration/images/utils/getPublicPath.ts)
- relative paths are explicit, starting with `../` or `./`
- and remote images are downloaded ... 😱 oh god I realised I didn't finish this 😵‍💫 just noticed an image broken in prod!!! 

It is certainly be possible to also have "absolute" paths or "public/" based paths. If I ever need those I'll probably use an alias `@public`.

### More details 

- [MDX authoring experience in Astro / Shortcomings](https://andretorgal.com/posts/2023-01/mdx-authoring-experience-in-astro-shortcomings-and-astro-m2dx/)
- [Improving (MDX) authoring experience in Astro / Solutions](https://andretorgal.com/posts/2023-01/improving-mdx-experience-in-astro-with-relative-paths-responsive-images-and-more/)

TODOs: 
- Add a circuit breaker for paths that land outside of the project (an allow list of paths would also work) 
- Opt-in - on a per image basis - to extracting a tiny blob `data:image/png;base64`

**Comment by Princesseuh** [comment]
Hi everyone! We're currently working towards a new core story for optimized images, including support for relative images in Markdown.

Please leave your feedback here: https://github.com/withastro/roadmap/discussions/468

**Comment by wassfila** [comment]
I updated the Survey section of this Discussion to take https://github.com/withastro/roadmap/discussions/468 into account. With a bit of luck, we might not need this one anymore, but we're not there yet.

**Comment by Princesseuh** [comment]
This is now supported by default in Astro 3.0
